File: python/81-90/Maximal-Rectangle.py
    def largestRectangleArea(self, height):
        if len(height) ==0:
            return 0
        height.append(0)
        stack = []
        mx = 0
        for i in xrange(len(height)):
            while len(stack) >0 and height[stack[-1]]>height[i]:
                h = height[stack[-1]]
                del stack[-1]
                left = -1
                if len(stack)>0:
                    left = stack[-1]
                if (h*(i-1-left)>mx):
                    mx = h*(i-1-left)
            stack.append(i)
        return mx
    # @param {character[][]} matrix
    # @return {integer}
    def maximalRectangle(self, matrix):
        if matrix == None or len(matrix)==0 or len(matrix[0])==0:
            return 0
        row,col = len(matrix),len(matrix[0])
        mx =0
        line = [0 for i in xrange(col)]
        for i in xrange(row):
            for j in xrange(col):
                if matrix[i][j]=='1':
                    line[j] += 1
                else:
                    line[j]=0
            tmx = self.largestRectangleArea(line[:])
            if tmx > mx:
                mx =tmx
        return mx

File: Cpp/81-90/Maximal Rectangle.cpp
/*
    runtime:28ms
    error: 0

*/
class Solution {
public:
        int largestRectangleArea(vector<int>& height) {
        stack<int> myStack;
        if(height.size()==0) return 0;
        height.push_back(0);
        int Max=0;
        for(int i=0; i<height.size(); i++){
            while(!myStack.empty()&&height[i]<height[myStack.top()]){
                int h=height[myStack.top()];
                myStack.pop();
                int wid = myStack.empty()?-1:myStack.top();
                int Area = (i-1-wid)*h;
                if(Area>Max) Max = Area;
            }
            myStack.push(i);
        }
        return Max;
    }
    int maximalRectangle(vector<vector<char>>& matrix) {
        if(matrix.size()==0||matrix[0].size()==0) return 0;
        int row = matrix.size();
        int col = matrix[0].size();
        for(int i=0; i<row; i++)
            matrix[i].push_back(0);
        int Max=0; vector<int> height(col+1,0);
        for(int i=0; i<row; i++){
            for(int j=0; j<col+1; j++)
                height[j] = (matrix[i][j]=='1')?height[j]+1:0;
            int Area = largestRectangleArea(height);
            Max = max(Max,Area);
        }
     return Max;   
    }
};
